### PR TITLE
[addons] add function for removing orphaned dependencies recursively

### PR DESCRIPTION
**Pull request by howie-f**
## Description

**Commit 1**
add-ons cleanup: replace bool default argument with enum class. the underlying data type for existing enum classes is changed to bool where appropiate

**Commit 2**
let's remove orphaned dependencies recursively. function located at `Settings -> System -> Add-ons`

**Commit 3**
uninstall of an add-on now removes any orphaned dependencies by default

## How has this been tested?
debian buster - tested with install/uninstall of skin `Aeon Nox: SILVO` which leaves several depends in an orphaned state

## What is the effect on users?
ability to batch clean orphaned dependencies

## Screenshots (if appropriate):
screenshot00000

screenshot00001

screenshot00002

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [X] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document

maybe @phunkyfish has time / mood for a review?

### CODE REVIEW

**Comment by howie-f** [comment]
the latest commit triggers the cleaning feature after a regular add-on uninstall. result in my log during testing:
```
INFO <general>: CAddonUnInstallJob[skin.aeon.nox.silvo]:
     removed orphaned dependencies (Skin Shortcuts, Simple Eval, unidecode)
```

File: xbmc/addons/AddonManager.cpp [hunk 1]
```diff
@@ -22,6 +22,7 @@
 #include "filesystem/Directory.h"
 #include "filesystem/File.h"
 #include "filesystem/SpecialProtocol.h"
+#include "messaging/helpers/DialogOKHelper.h"
```
Comment (ksooo) [inline xbmc/addons/AddonManager.cpp:25]: I don't know the architecture of the addon subsystem, but it seems you are polluting a core component with GUI stuff here, which feels like bad design.

File: addons/resource.language.en_gb/resources/strings.po [hunk 2]
```diff
@@ -15811,7 +15811,10 @@ msgctxt "#24162"
 msgid "Use Kodi virtual keyboard"
 msgstr ""
 
-#empty string with id 24163
+#: ./system/settings/settings.xml
+msgctxt "#24163"
+msgid "Remove orphaned dependencies recursively"
```
Comment (ksooo) [inline addons/resource.language.en_gb/resources/strings.po:15816]: Is this correct English or should the read "Recursively remove ..." ? And is the term "recursively" something a normal user understands or is it by far to technical for John Doe? I would phrase this "Remove all orphaned dependencies"

File: addons/resource.language.en_gb/resources/strings.po [hunk 3]
```diff
@@ -21110,7 +21113,28 @@ msgctxt "#36637"
...
+msgstr ""
+
+#: xbmc/addons/AddonSystemSettings.cpp
+msgctxt "#36640"
+msgid "The following orphaned dependency add-ons were removed from your system: {0:s}."
+msgstr ""
+
+#: xbmc/addons/AddonSystemSettings.cpp
+msgctxt "#36641"
+msgid "No orphaned dependency add-ons found."
```
Comment (ksooo) [inline addons/resource.language.en_gb/resources/strings.po:21134]: This reads strange. Do you mean "No add-ons with orphaned dependencies found." ?

File: addons/resource.language.en_gb/resources/strings.po [hunk 4]
```diff
@@ -21110,7 +21113,28 @@ msgctxt "#36637"
...
 msgstr ""
 
-#empty strings from id 36638 to 36898
+#. Description of setting with label #24163 "Remove orphaned dependencies recursively"
+#: system/settings/settings.xml
+msgctxt "#36638"
+msgid "Remove all modules and support libraries that have been automatically installed as a dependency to other add-ons and are in an orphaned state."
+msgstr ""
+
+#: xbmc/addons/AddonSystemSettings.cpp
```
Comment (ksooo) [inline addons/resource.language.en_gb/resources/strings.po:21122]: Please add a meaningful "#."  description - here and for the other new strings.

File: xbmc/addons/AddonInstaller.h [hunk 5]
```diff
@@ -270,15 +270,19 @@ class CAddonInstallJob : public CFileOperationJob
 class CAddonUnInstallJob : public CFileOperationJob
 {
 public:
-  CAddonUnInstallJob(const ADDON::AddonPtr &addon, bool removeData);
+  CAddonUnInstallJob(const ADDON::AddonPtr& addon,
+                     bool removeData,
+                     bool removeOrphanedDependencies);
```
Comment (ksooo) [inline xbmc/addons/AddonInstaller.h:275]: Argh. Another new bool param which leads to unreadable code at the calling side.

File: xbmc/addons/AddonManager.cpp [hunk 6]
```diff
@@ -297,29 +298,90 @@ bool CAddonMgr::HasAvailableUpdates()
...
+  std::vector<std::string> removedItems;
+
+  for (auto toRemove = GetOrphanedDependencies(); toRemove.size() > 0;
+       toRemove = GetOrphanedDependencies())
+  {
+    for (const auto& dep : toRemove)
+    {
+      bool removeData = CDirectory::Exists("special://profile/addon_data/" + dep->ID());
+      CAddonUnInstallJob removeDependency(dep, removeData, false);
+      if (removeDependency.DoWork())
```
Comment (ksooo) [inline xbmc/addons/AddonManager.cpp:347]: Oh really? Misusing a class designed for async executing by directly calling its worker method? This at least looks like a hack, even though it works.

File: xbmc/addons/AddonManager.cpp [hunk 7]
```diff
@@ -297,29 +298,90 @@ bool CAddonMgr::HasAvailableUpdates()
...
+  }
+  return true;
+}
+
+std::vector<std::string> CAddonMgr::RemoveOrphanedDependencies() const
+{
+  std::vector<std::string> removedItems;
+
+  for (auto toRemove = GetOrphanedDependencies(); toRemove.size() > 0;
+       toRemove = GetOrphanedDependencies())
```
Comment (ksooo) [inline xbmc/addons/AddonManager.cpp:341]: This looks strange. Could this be expressed shorter/more elegant with a `while` loop?

File: xbmc/addons/AddonManager.h [hunk 8]
```diff
@@ -154,6 +160,16 @@ namespace ADDON
     /*! Returns true if there is any addon with available updates, otherwise false */
     bool HasAvailableUpdates();
 
+    /*! Returns true if the passed addon is an orphaned dependency, otherwise false */
```
Comment (ksooo) [inline xbmc/addons/AddonManager.h:163]: Please add proper doxy.

File: xbmc/addons/AddonManager.h [hunk 9]
```diff
@@ -557,8 +573,11 @@ namespace ADDON
 
     bool GetAddonsInternal(const TYPE& type,
                            VECADDONS& addons,
-                           bool onlyEnabled,
-                           bool checkIncompatible = false) const;
+                           OnlyEnabled onlyEnabled,
+                           CheckIncompatible checkIncompatible) const;
+
+    /*! Get all dependency add-ons in an orphaned state */
```
Comment (ksooo) [inline xbmc/addons/AddonManager.h:579]: Please add proper doxy.

File: xbmc/addons/gui/GUIDialogAddonInfo.cpp [hunk 10]
```diff
@@ -555,7 +555,7 @@ void CGUIDialogAddonInfo::OnUninstall()
   if (CDirectory::Exists("special://profile/addon_data/" + m_localAddon->ID()))
     removeData = CGUIDialogYesNo::ShowAndGetInput(CVariant{24037}, CVariant{39014});
 
-  CJobManager::GetInstance().AddJob(new CAddonUnInstallJob(m_localAddon, removeData),
+  CJobManager::GetInstance().AddJob(new CAddonUnInstallJob(m_localAddon, removeData, true),
```
Comment (ksooo) [inline xbmc/addons/gui/GUIDialogAddonInfo.cpp:558]: This is the unreadable code I meant when moaning about new bool params... I need to look at the method declaration to understand what the `true` means.

File: xbmc/platform/linux/input/LibInputSettings.cpp [hunk 11]
```diff
@@ -58,7 +58,7 @@ CLibInputSettings::CLibInputSettings(CLibInputHandler *handler) :
   std::set<std::string> settingSet;
   settingSet.insert(SETTING_INPUT_LIBINPUTKEYBOARDLAYOUT);
   settingsManager->RegisterCallback(this, settingSet);
-  settingsManager->RegisterSettingOptionsFiller("libinputkeyboardlayout",
+  settingsManager->RegisterSettingOptionsFiller("input.libinputkeyboardlayout",
```
Comment (ksooo) [inline xbmc/platform/linux/input/LibInputSettings.cpp:61]: This change looks unrelated to this PR.

File: xbmc/platform/linux/input/LibInputSettings.cpp [hunk 12]
```diff
@@ -142,7 +142,7 @@ CLibInputSettings::~CLibInputSettings()
   if (!settings)
     return;
 
-  settings->GetSettingsManager()->UnregisterSettingOptionsFiller("libinputkeyboardlayout");
+  settings->GetSettingsManager()->UnregisterSettingOptionsFiller("input.libinputkeyboardlayout");
```
Comment (ksooo) [inline xbmc/platform/linux/input/LibInputSettings.cpp:145]: This change looks unrelated to this PR.

File: addons/resource.language.en_gb/resources/strings.po (hunk 2 above)
Comment (howie-f) [reply addons/resource.language.en_gb/resources/strings.po:15816]: yes we can call it `Remove all orphaned dependencies`.
point is after my test - install/uninstall aeon.silvo - you see only **1** orphaned at `Settings -> System -> Add-ons - Manage dependencies`.... only `script.skinshortcuts`,... and after that one is removed the next 2 orphaned are revealed. but you're right, this is not important to John

File: addons/resource.language.en_gb/resources/strings.po (hunk 3 above)
Comment (howie-f) [reply addons/resource.language.en_gb/resources/strings.po:21134]: no it should say: there are no more orphaned add-ons left that i could delete

File: xbmc/addons/AddonInstaller.h (hunk 5 above)
Comment (howie-f) [reply xbmc/addons/AddonInstaller.h:275]: thought the same exactly, will switch to an enum

File: xbmc/addons/AddonManager.cpp (hunk 7 above)
Comment (howie-f) [reply xbmc/addons/AddonManager.cpp:341]: i had a `do...while` in the first place but still needed to initialize `toRemove`. let's see if i find something prettier

File: xbmc/addons/AddonManager.cpp (hunk 6 above)
Comment (howie-f) [reply xbmc/addons/AddonManager.cpp:347]: yup, looks like a hack to me too. but i don't know better atm...

File: xbmc/addons/gui/GUIDialogAddonInfo.cpp (hunk 10 above)
Comment (howie-f) [reply xbmc/addons/gui/GUIDialogAddonInfo.cpp:558]: will be enum class for sure

File: xbmc/platform/linux/input/LibInputSettings.cpp (hunk 11 above)
Comment (howie-f) [reply xbmc/platform/linux/input/LibInputSettings.cpp:61]: that was a coincidence catch while diggin in the log, open a separate pr for this? ok

File: xbmc/addons/AddonManager.cpp (hunk 1 above)
Comment (howie-f) [reply xbmc/addons/AddonManager.cpp:25]: ok, need to shift the dialog somewhere else..

Comment (howie-f) [reply xbmc/addons/AddonManager.cpp:25]: damn, that was a leftover, the dialog already moved to ` xbmc/addons/AddonSystemSettings.cpp`

File: xbmc/platform/linux/input/LibInputSettings.cpp (hunk 11 above)
Comment (howie-f) [reply xbmc/platform/linux/input/LibInputSettings.cpp:61]: @lrusak please feel free to pick this, so i can throw it outta here

File: addons/resource.language.en_gb/resources/strings.po (hunk 3 above)
Comment (phunkyfish) [reply addons/resource.language.en_gb/resources/strings.po:21134]: So: `No orphaned dependency add-ons to delete`?

File: xbmc/addons/AddonManager.cpp (hunk 7 above)
Comment (phunkyfish) [reply xbmc/addons/AddonManager.cpp:341]:
```cpp
auto toRemove = GetOrphanedDependencies();
while (toRemove.size() > 0)
  toRemove = GetOrphanedDependencies())
```

File: addons/resource.language.en_gb/resources/strings.po (hunk 2 above)
Comment (howie-f) [reply addons/resource.language.en_gb/resources/strings.po:15816]: done

File: addons/resource.language.en_gb/resources/strings.po (hunk 4 above)
Comment (howie-f) [reply addons/resource.language.en_gb/resources/strings.po:21122]: done

File: addons/resource.language.en_gb/resources/strings.po (hunk 3 above)
Comment (howie-f) [reply addons/resource.language.en_gb/resources/strings.po:21134]: done

File: xbmc/addons/AddonInstaller.h (hunk 5 above)
Comment (howie-f) [reply xbmc/addons/AddonInstaller.h:275]: done

File: xbmc/addons/AddonManager.cpp (hunk 7 above)
Comment (howie-f) [reply xbmc/addons/AddonManager.cpp:341]: thanks, done

File: xbmc/addons/AddonManager.h (hunk 8 above)
Comment (howie-f) [reply xbmc/addons/AddonManager.h:163]: done

File: xbmc/addons/AddonManager.h (hunk 9 above)
Comment (howie-f) [reply xbmc/addons/AddonManager.h:579]: done

File: xbmc/addons/gui/GUIDialogAddonInfo.cpp (hunk 10 above)
Comment (howie-f) [reply xbmc/addons/gui/GUIDialogAddonInfo.cpp:558]: done

File: xbmc/addons/AddonManager.cpp (hunk 6 above)
Comment (howie-f) [reply xbmc/addons/AddonManager.cpp]: hum, still feels like a hack. but we're doing it similarly since Isengard in
```c
bool CAddonInstallJob::Install(const std::string &installFrom, const AddonPtr& repo)
```
https://github.com/xbmc/xbmc/blob/02e7013889e08d608363f2909ebeccdb9ea3b7c9/xbmc/addons/AddonInstaller.cpp#L807-L813

best bet would be changing the direct call to`DoWork()` into `DoModal()` with no progress indicator

**Comment by phunkyfish** [comment]
Is this ready for review now @howie-f ?

**Comment by howie-f** [comment]
> Is this ready for review now

yes, i‘ve nothing to change at the moment

File: xbmc/addons/AddonManager.cpp (hunk 6 above)
Comment (phunkyfish) [reply xbmc/addons/AddonManager.cpp]: I'm not sure what the impact of using DoModal() instead is. Not what a correct alternative would be. @ksooo would you suggest something different?

Comment (howie-f) [reply xbmc/addons/AddonManager.cpp]: sorry guys, i now see this uninstall logic better fits in `CAddonInstaller` singleton, not `CAddonMgr`... but the main question DoWork(), DoModal() or something completely different remains

File: xbmc/addons/AddonInstaller.cpp [hunk 13]
```diff
@@ -207,6 +207,15 @@ bool CAddonInstaller::InstallOrUpdateDependency(const ADDON::AddonPtr& dependsId
...
 }
 
+bool CAddonInstaller::RemoveDependency(const std::shared_ptr<IAddon>& dependsId) const
+{
+  bool removeData = CDirectory::Exists("special://profile/addon_data/" + dependsId->ID());
+  CAddonUnInstallJob removeDependencyJob(dependsId, removeData);
+  removeDependencyJob.SetRecurseOrphaned(RecurseOrphaned::NO);
+
+  return removeDependencyJob.DoWork();
+}
```
Comment (howie-f) [inline xbmc/addons/AddonInstaller.cpp:217]: imo `removeDependencyJob.DoWork()` is correct here as we should not show a progress indicator. we cannot know how many deps will get purged beforehand.

**Review by howie-f — COMMENTED** [review]
`CAddonInstaller::RemoveDependency` added... was misplaced in AddonManager

File: xbmc/addons/AddonInstaller.cpp (hunk 13 above)
Comment (howie-f) [reply xbmc/addons/AddonInstaller.cpp:217]: @Montellese what‘s your opinion on this? i think this section was your work long time ago

**Comment by howie-f** [comment]
jenkins build this please

**Comment by howie-f** [comment]
@ksooo and @phunkyfish may i ask for a final review? i'm convinced the direct call to `removeDependencyJob.DoWork()` is ok. we're doing the same thing when installing dependencies.

**Comment by howie-f** [comment]
just a rebase, nothing changed

File: xbmc/addons/AddonInstaller.h [hunk 14]
```diff
@@ -26,40 +26,46 @@ namespace ADDON
 
 class CAddonDatabase;
 
-enum class BackgroundJob
+enum class BackgroundJob : bool
```
Comment (phunkyfish) [inline xbmc/addons/AddonInstaller.h:29]: If the enum will essentially be a bool might make more sense to have the enum values being TRUE and FALSE!

**Review by phunkyfish — COMMENTED** [review]
Just a comment, up to you but might be easier to understand and less leaving it open to interpretation. Other than that it all looks reasonable.

File: xbmc/addons/AddonInstaller.h (hunk 14 above)
Comment (howie-f) [reply xbmc/addons/AddonInstaller.h:29]: that dont work because we have a name clash then with macros TRUE/FALSE.. was in the very beginning of our add-on system refactoring. yeah i do remember 🙂

**Review by phunkyfish — APPROVED** [review]
Looks sane to me.

File: xbmc/addons/AddonInstaller.h (hunk 14 above)
Comment (phunkyfish) [reply xbmc/addons/AddonInstaller.h:29]: Everything comes full circle! Now you’re mentoring me!!!

Comment (howie-f) [reply xbmc/addons/AddonInstaller.h:29]: phunky, not really 😅 i know it was just a test that i passed now.. haha

**Review by enen92 — APPROVED** [review]
